core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -226,48 +357,215 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
-      val messages = this.internalIterator(isShallow = false).map(messageAndOffset => {
-        if (compactedTopic && !messageAndOffset.message.hasKey)
-          throw new InvalidMessageException("Compacted topic cannot accept message without key.")
-
-        messageAndOffset.message
+      // Deal with compressed messages
+      // We only need to do re-compression in one of the followings situation:
+      // 1. Source and target compression codec are different
+      // 2. When magic value to use is 0 because offsets need to be overwritten
+      // 3. When magic value to use is above 0, but some fields of inner messages need to be overwritten.
+      // 4. Message format conversion is needed.
+
+      // Re-compression situation 1 and 2
+      var requireReCompression = sourceCodec != targetCodec || magicValueToUse == Message.MagicValue_V0
+
+      var maxTimestamp = Message.NoTimestamp
+      val expectedRelativeOffset = new AtomicLong(0)
+      val validatedMessages = new ListBuffer[Message]
+      this.internalIterator(isShallow = false).foreach(messageAndOffset => {
+        val message = messageAndOffset.message
+        validateMessageKey(message, compactedTopic)
+        if (message.magic > Message.MagicValue_V0 && magicValueToUse > Message.MagicValue_V0) {
+          // Re-compression situation 3
+          // Validate the timestamp
+          validateTimestamp(message, now, messageTimestampType, messageTimestampDiffMaxMs)
+          // Check if we need to overwrite offset
+          if (messageAndOffset.offset != expectedRelativeOffset.getAndIncrement)
+            requireReCompression = true
+          maxTimestamp = math.max(maxTimestamp, message.timestamp)
+        }
+
+        // Re-compression situation 4
+        if (message.magic != magicValueToUse)
+          requireReCompression = true
+
+        validatedMessages += message.toFormatVersion(magicValueToUse)
       })
-      new ByteBufferMessageSet(compressionCodec = targetCodec, offsetCounter = offsetCounter, messages = messages.toBuffer:_*)
+
+      if (requireReCompression) {
+        // Re-compression required.
+        val messageSetTimestampAssignor = (messages: Seq[Message]) => {
+          if (magicValueToUse == Message.MagicValue_V0)
+            Message.NoTimestamp
+          else if (magicValueToUse > Message.MagicValue_V0 && messageTimestampType == TimestampType.CreateTime)
+            maxTimestamp
+          else // Log append time
+            now
+        }
+
+        new ByteBufferMessageSet(compressionCodec = targetCodec,
+                                 offsetCounter = offsetCounter,
+                                 messageSetTimestampAssignor = messageSetTimestampAssignor,
+                                 timestampType = messageTimestampType,
+                                 messages = validatedMessages.toBuffer: _*)
+      } else {
```

Comment:
Passing in messageSetTimestampAssignor makes the code a bit harder to read. I was wondering if we can obviate that and always scan the messages to get the max timestamp. This will add a bit overhead of making another iteration of all messages, but will simplify the code. The overhead should be small.